### PR TITLE
[WIP/RFC] make-derivation: default name if pname and version are both set

### PR DESCRIPTION
**Pull request by Synthetica9**
###### Motivation for this change

This is done in, for example, `buildPythonPackage`

###### TODO

- [x] The basic change (c313e07) 
- [x] Remove `name = "${pname}-${version}"` from packages that define it that way (ef6ece0) 
- [ ] Update a lot of Gnome packages such that they don't use `${name}` in the `fetchUrl` anymore (see
af1b4f4)
- [ ] Change `name = "foobar-${version}"` for to `pname = "foobar"` for packages that define it that way (WIP on my machine, will give a lot of pings once I push it: there are 3k+ packages that use this method.)
- [ ] Documentation:
    - [ ] Document that this exists
    - [ ] Recommend it as the default way of defining packages?
- [ ] Cleanup:
    - [ ] Remove `rec` from packages that don't need it anymore
    - [ ] Squash/reorder a few commits in this pr
- [ ] A full build to check that I didn't make any mistakes.

---

### CODE REVIEW

**Comment by dtzWill** [comment]
I'm not sure how important it is but I think it's common for these to be `rec` only for this reason, making them non-recursive would be nice: simpler for human understanding, friendlier to eval at least in theory.

File: pkgs/stdenv/generic/make-derivation.nix [hunk 1]
```diff
@@ -12,7 +12,9 @@ rec {
   # * https://nixos.org/nix/manual/#ssec-derivation
   #   Explanation about derivations in general
   mkDerivation =
-    { name ? ""
+    { name ? if builtins.hasAttr "pname" attrs && builtins.hasAttr "version" attrs
```
Comment (matthewbauer) [inline pkgs/stdenv/generic/make-derivation.nix]: Can you update "pos" above as well? It is used to find line numbers in traces. I think we need to also look for "version" before name for this to work correctly.

**Review by matthewbauer — CHANGES_REQUESTED** [review]
Definitely looks like something very useful! pname/version is what lots of packages are already using so hopefully everyone is okay with this.

File: pkgs/stdenv/generic/make-derivation.nix (hunk 1 above)
Comment (Synthetica9) [reply pkgs/stdenv/generic/make-derivation.nix]: Would that be something like `builtins.unsafeGetAttrPos "pname" attrs`?

Comment (matthewbauer) [reply pkgs/stdenv/generic/make-derivation.nix]: Yeah but I actually think version would be better because it's more frequently used than pname.

Comment (Synthetica9) [reply pkgs/stdenv/generic/make-derivation.nix]: Okay, I'll add it.

Comment (Synthetica9) [reply pkgs/stdenv/generic/make-derivation.nix]: Does 0c1d5d1 look good to you?

**Review by matthewbauer — APPROVED** [review]

File: pkgs/stdenv/generic/make-derivation.nix (hunk 1 above)
Comment (Synthetica9) [reply pkgs/stdenv/generic/make-derivation.nix]: Oh, yeah, that would make sense. Adding that.

Comment (Synthetica9) [reply pkgs/stdenv/generic/make-derivation.nix]: @globin e0d2348

**Comment by matthewbauer** [comment]
@globin I suspect we should document this as well?

**Comment by matthewbauer** [comment]
Try removing pname & version from attrs in python here:

https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/python/mk-python-derivation.nix#L68

**Comment by Synthetica9** [comment]
@matthewbauer Looks like a good first step, but crashes here (and someone foresaw it :P):

https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/setuptools/default.nix#L8

**Comment by matthewbauer** [comment]
Yeah we might be able to just remove the prefix there as well.

/cc @FRidh

**Comment by FRidh** [comment]
I'm of the opinion version should not actually be an attribute of the derivation but instead of the source.

**Comment by matthewbauer** [comment]
>I'm of the opinion version should not actually be an attribute of the derivation but instead of the source.

That's a good point - but the way Nixpkgs works now is that version is an attribute of the derivation through name - this just lets you separate those two out.

**Comment by FRidh** [comment]
> That's a good point - but the way Nixpkgs works now is that version is an attribute of the derivation through name - this just lets you separate those two out.

We include it in the name, but it is not an attribute. Therefore, it is possible to have `name = "${pname}-${src.version}`.

**Comment by FRidh** [comment]
I've pushed b5d7bb429204c2f420628a4d49fe5bdd9cc214bc (removing name from Python packages) to staging.

**Comment by FRidh** [comment]
> Try removing pname & version from attrs in python here:

> Looks like a good first step, but crashes here (and someone foresaw it :P):

In the case of Python, you could set
```
pname = "${python.libPrefix}-${pname}"
```
and remove the `name` and `version` parameters.

**Comment by Synthetica9** [comment]
@FRidh 

The following should also work I think:

```nix
let
  pname = "setuptools";
  version = "39.0.1";
in stdenv.mkDerivation rec {
  name = "${python.libPrefix}-${pname}-${version}";
```

**Comment by Synthetica9** [comment]
I think I'm starting to come to the conclusion that dropping pname and/or version from pythonpackages is just a bad idea ([people find creative ways to use both](https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/python-packages.nix#L10128)), and we should just change the assert to `hasSuffix name "${pname}-${version}"`

**Comment by FRidh** [comment]
> I think I'm starting to come to the conclusion that dropping pname and/or version from pythonpackages is just a bad idea (people find creative ways to use both), and we should just change the assert to hasSuffix name "${pname}-${version}"

You are referring to the following code:
```nix
  psutil_1 = self.psutil.overrideAttrs (oldAttrs: rec {
    name = "${oldAttrs.pname}-${version}";
    version = "1.2.1";
    src = oldAttrs.src.override {
      inherit version;
      sha256 = "0ibclqy6a4qmkjhlk3g8jhpvnk0v9aywknc61xm3hfi5r124m3jh";
    };
});
```
Here they need to set name because they use `overrideAttrs` which overrides the call to `stdenv.mkDerivation`, and creation of name is done in `buildPythonPackage`. Instead, if `overridePythonAttrs` is used, then there would not have been any need for setting `name`.

With your proposed change, `stdenv.mkDerivation` will fix `name`, so then they technically could use `overrideAttrs` and without passing in a custom `name`. So, it's not that dramatic at all. 

If you like to I can have a look at how `buildPython*` needs to be changed. Tomorrow I should have time for it.

**Comment by edolstra** [comment]
A fundamental change like this is something that should be done via the RFC process, IMHO.

**Comment by Synthetica9** [comment]
Superseded by https://github.com/NixOS/rfcs/pull/35

File: pkgs/stdenv/generic/make-derivation.nix (hunk 1 above)
Comment (edolstra) [reply pkgs/stdenv/generic/make-derivation.nix]: This should be `attrs ? pname && attrs ? version`.

**Pull request closed without merging**